### PR TITLE
Intermediate

### PR DESCRIPTION
**Pull request by khamraeva**

### CODE REVIEW

File: frontend/app/page-size/page-size.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  });
+
+  it('should display first option by default', () => {
+    const elements = fixture.debugElement.queryAll(By.css('[test-id="page-size-options"]'));
+    const firstOption = elements[0];
+    expect(firstOption.nativeElement.selected).toBe(true, 'Incorrect default option')
+  });
+
+  it('should be clicked', () => {
+    component.selectChangeHandler(2);
```
Comment (strafe89) [inline frontend/app/page-size/page-size.component.spec.ts:42]: !

File: README.md [hunk 2]
```diff
@@ -58,3 +58,7 @@ yarn start
 ```
 yarn test
 ```
+
+EXERCISE NOTES
```
Comment (strafe89) [inline README.md:62]: !

File: frontend/app/app.component.spec.ts [hunk 3]
```diff
@@ -112,26 +120,26 @@ describe('AppComponent', () => {
     });
 
     it('should navigate to the first page when the provided page is < 1', async () => {
-      setUpOrderFixturesForEndpoint('http://localhost:4300/orders?page=1', {
+      setUpOrderFixturesForEndpoint('http://localhost:4300/orders?page=1&pageSize=10', {
```
Comment (strafe89) [inline frontend/app/app.component.spec.ts:123]: mocked response should be aligned with request parameters.
example: page size in request should be equal to page size in response

File: frontend/app/app.module.ts [hunk 4]
```diff
@@ -7,6 +7,11 @@ import { OrderRowComponent } from './order-row/order-row.component';
 import { StatusPillComponent } from './status-pill/status-pill.component';
 import { RouterModule, Routes } from '@angular/router';
 import { HomepageComponent } from './homepage/homepage.component';
+import { HttpService } from './services/httpService.service';
+import { PaginationComponent } from './pagination/pagination.component';
+import { PaginationService } from './services/pagination.service';
+import { PageSizeComponent } from './page-size/page-size.component';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
```
Comment (strafe89) [inline frontend/app/app.module.ts:14]: nit: external libraries imports should go first

File: frontend/app/homepage/homepage.component.spec.ts [hunk 5]
```diff
@@ -1,14 +1,108 @@
 import { ComponentFixture, TestBed } from '@angular/core/testing';
 import { HomepageComponent } from './homepage.component';
+import { AppModule } from '../app.module';
+import { By } from '@angular/platform-browser';
+import { HttpService } from '../services/httpService.service';
+import { of } from 'rxjs';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
 
 describe('HomepageComponent', () => {
+  const mockData = {
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.spec.ts:10]: data mock can be moved to separate json file

File: frontend/app/page-size/page-size.component.html [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<div style="margin:10px;">
```
Comment (strafe89) [inline frontend/app/page-size/page-size.component.html:1]: inlined styles = bad pattern?

File: frontend/app/pagination/pagination.component.html [hunk 7]
```diff
@@ -0,0 +1,35 @@
+<nav>
+  <ul class="pagination">
+    <li
+      class="page-item"
+      [class.disabled]="currentPage === 1">
+      <button
+        class="page-link"
+        test-id="prev-page-nav-item"
+        (click)="currentPage !== 1 && getOrdersByPage(currentPage - 1)">
```
Comment (strafe89) [inline frontend/app/pagination/pagination.component.html:9]: pagination component shouldn't know anything about specific entities (orders)

File: package.json [hunk 8]
```diff
@@ -24,6 +24,7 @@
     "concurrently": "^3.5.1",
     "express": "^4.16.3",
     "moment": "^2.22.1",
+    "ngx-pagination": "^6.0.3",
```
Comment (strafe89) [inline package.json:27]: where this library is used in the code?

File: frontend/app/services/pagination.service.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
+import { Injectable } from '@angular/core';
+
+@Injectable()
+export class PaginationService {
+
+  setPaginationRange(currentPage: number, totalPages: number): number[] {
+    let range = 5,
```
Comment (strafe89) [inline frontend/app/services/pagination.service.ts:7]: range and pagesRange can be consts

File: frontend/app/services/httpService.service.spec.ts [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+        vendor: 'Mando\'s',
+        commissionRate: 0.15,
+        requestedDeliveryDate: '2018-04-09T06:30:00.000Z',
+        price: {
+          delivery: 0,
+          items: 5299.75,
+          total: 5299.75,
+          vatRate: 20,
+          vatableItems: 0,
+          vatAmount: 0},
```
Comment (strafe89) [inline frontend/app/services/httpService.service.spec.ts:23]: formatting

File: frontend/app/pagination/pagination.component.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+  @Output() pageChanged: EventEmitter<number> = new EventEmitter<number>();
+  paginationRange: number[] = [];
+
+  constructor(private paginationService: PaginationService) {}
+
+  ngOnChanges() {
+    this.paginationRange = this.paginationService.setPaginationRange(this.currentPage, this.totalPages);
+  }
+
+  getOrdersByPage(selectedPageNumber: number): void {
```
Comment (strafe89) [inline frontend/app/pagination/pagination.component.ts:29]: again, pagination component should be abstract and not tied to specific entities (orders)

File: frontend/app/homepage/homepage.component.ts [hunk 12]
```diff
@@ -1,33 +1,61 @@
...
+    this.httpService.getData(changedPage, pageSize)
+    .subscribe(res => {
+      this.orders = res.items.map(item => createOrderFromJsonDTO(item));
+      this.pageSize = res.pageSize;
+      this.totalPages = Math.floor(res.total / this.pageSize);
+      this.currentPage = res.page;
+    });
+    
+    if(this.currentPage > this.totalPages) {
+      this.getPageData(this.initialPage, pageSize)
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.ts:42]: ;

File: frontend/app/homepage/homepage.component.ts [hunk 13]
```diff
@@ -1,33 +1,61 @@
...
+    };
+
+    this.router.navigate(['/'], {
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.ts:45]: according to its name method "get page data" should be responsible for data fetching, why it also performs a redirect?

File: frontend/app/homepage/homepage.component.spec.ts [hunk 14]
```diff
@@ -17,11 +111,27 @@ describe('HomepageComponent', () => {
...
-  // Note to candidates: This method should pass as part of task 1
   it('should create', () => {
     expect(component).toBeTruthy();
   });
+
+  it('should call getPageData', () => {
+    expect(component.getPageData).toHaveBeenCalled();
+  });
+
+  it('should contain page size options', () => {
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.spec.ts:127]: not sure it's right that container (homepage) component tests contains tests of it's child component (pagination)

File: frontend/app/homepage/homepage.component.ts [hunk 15]
```diff
@@ -1,33 +1,61 @@
...
-    this.orders$ = this.httpClient.get(`${environment.apiBase}/orders?page=1`).pipe(
-      map((json: any) => json.items.map(createOrderFromJsonDTO))
-    );
-  }
-}
+  public constructor (
+    private httpService: HttpService,
+    private activatedRoute: ActivatedRoute,
+    private router: Router) {
+      const queryPage = Number(this.activatedRoute.snapshot.queryParams.page);
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.ts:23]: move queryParams to the const
```
const { queryParams } = this.activatedRoute.snapshot;
```

File: frontend/app/homepage/homepage.component.spec.ts [hunk 16]
```diff
@@ -17,11 +111,27 @@ describe('HomepageComponent', () => {
...
+    const el = fixture.debugElement.query(By.css('app-page-size'));
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.spec.ts:128]: some tests query html elements by "test id", some query by CSS selectors. that's not consistent

File: frontend/app/services/pagination.service.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+
+    if (currentPage < (range / 2) + 1 ) {
+      start = 1;
+    } else if (currentPage >= (totalPages - (range / 2) )) {
+      start = Math.floor(totalPages - range + 1);
+    } else {
+      start = (currentPage - Math.floor(range / 2));
+    }
+
+    for (let i = start; i <= ((start + range) - 1); i++) {
```
Comment (strafe89) [inline frontend/app/services/pagination.service.ts:19]: for (let i = start; i < start + range; i++) {

File: frontend/app/services/httpService.service.ts [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+import { Injectable } from '@angular/core';
+import { Page } from '../types/page';
+import { OrderDTO } from '../types/order';
+
+@Injectable()
+export class HttpService {
+
+  constructor(private httpClient: HttpClient) {}
+
+  getData(page: number, pageSize: number): Observable<Page<OrderDTO>> {
```
Comment (strafe89) [inline frontend/app/services/httpService.service.ts:13]: i'd suggest renaming this method to "getOrders" to be more specific. 
`httpService.getData` - currently it's not clear at all, which data it's going to fetch

**Comment by khamraeva** [comment]
Response contains page DTO as well, not only orders

пт, 30 дек. 2022 г. в 11:44, Gulkovskiy Dmitry ***@***.***>:

File: frontend/app/services/httpService.service.spec.ts [hunk 19]
```diff
@@ -0,0 +1,121 @@
...
+        ]
+    });
+  });
+
+  it('should get all orders', () => {
+    httpService = TestBed.inject(HttpService),
+    httpTestingController = TestBed.inject(HttpTestingController);
+    httpService.getData(1, 10)
+      .subscribe(orders => {
+        expect(orders).toBeTruthy('No orders returned');
```
Comment (strafe89) [inline frontend/app/services/httpService.service.spec.ts:104]: you can also check length of this array

File: frontend/app/services/httpService.service.spec.ts [hunk 20]
```diff
@@ -0,0 +1,121 @@
...
+      expect(req.request.method).toEqual('GET');
+
+      req.flush(mockData);
+
+  });
+
+  afterEach(() => {
+    httpTestingController.verify();
+  });
+});
```
Comment (strafe89) [inline frontend/app/services/httpService.service.spec.ts:121]: empty line missing

File: frontend/app/pagination/pagination.component.spec.ts [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+    component.paginationRange = range;
+    spyOn(component, 'getOrdersByPage');
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
+    expect(component).toBeTruthy();
+  });
+
+  it('should display correct pages', () => {
```
Comment (strafe89) [inline frontend/app/pagination/pagination.component.spec.ts:32]: "it should display correct pages" - but this test checks the content only of the first element.
i'd suggest to align test description with it's content

File: frontend/app/pagination/pagination.component.spec.ts [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+    const firstElement = Number(elements[0].nativeElement.textContent);
+    expect(firstElement).toBe(range[0], 'Incorrect pages were display');
+  });
+
+  it('should display correct amount of pages', () => {
+    const elements = fixture.debugElement.queryAll(By.css('[test-id="pagination-range"]'));
+    expect(elements.length).toBe(range.length, 'Incorrect amount of pages');
+  });
+
+  it('prev page button should trigger getOrdersByPage function', () => {
```
Comment (strafe89) [inline frontend/app/pagination/pagination.component.spec.ts:43]: similar tests can be added for "next" and "page 1,2,3..." buttons

File: frontend/app/services/httpService.service.ts (hunk 18 above)
Comment (khamraeva) [reply frontend/app/services/httpService.service.ts:13]: Response contains page params, not only orders

Comment (strafe89) [reply frontend/app/services/httpService.service.ts:13]: it's just a pagination metadata related to orders response

File: frontend/app/app.component.spec.ts (hunk 3 above)
Comment (strafe89) [reply frontend/app/app.component.spec.ts:123]: some mocks are not fixed

File: frontend/app/services/httpService.service.ts (hunk 18 above)
Comment (strafe89) [reply frontend/app/services/httpService.service.ts:13]: not renamed

File: frontend/app/homepage/homepage.component.spec.ts [hunk 23]
```diff
@@ -1,14 +1,29 @@
 import { ComponentFixture, TestBed } from '@angular/core/testing';
 import { HomepageComponent } from './homepage.component';
+import { AppModule } from '../app.module';
+import { By } from '@angular/platform-browser';
+import { HttpService } from '../services/httpService.service';
+import { HttpClientTestingModule } from '@angular/common/http/testing';
```
Comment (strafe89) [inline frontend/app/homepage/homepage.component.spec.ts:6]: not sorted

File: frontend/app/pagination/pagination.component.spec.ts (hunk 22 above)
Comment (strafe89) [reply frontend/app/pagination/pagination.component.spec.ts:41]: test description not updated

**Pull request merged**